### DISCUSSION TITLE
Ideas on synchronizing design and development workflow

### DISCUSSION THREAD

**Discussion by zkwsk** [react-figma/react-figma#601] [Ideas]
What are peoples thoughts on the following use case for react-figma?

I find myself working on a design system that has been under development for more than 5 years. Over time the designer switched to different tools and a lot of components have been changed without the original design files being kept up to date. Now, the designer has left the project and we no longer own all of the design files, but we do have a well structured design system.

To make it easier for a new designer to lift this legacy I had the idea of looking into ways to import "production grade" React components into Figma. This research led me to react-figma, but unfortunately it seems to not quite do this in its current form.

I think this could be very beneficial in a situation like mine, but also in a day to day workflow. Once a design is implemented or updated in code, the designer can import that design directly into Figma for future revisions, rather than trying to play catch-up. Further, things like component breakdown, which I often find designers and developers disagreeing about which could easily lead to designers working with a different mental model of how everything is structured and broken down compared to what is actually implemented. I know it might be uncomfortable for some designers to "loose ownership" over the design, but it is still the designer who created the design, the only difference is that the "source of truth" is now in code rather than in some design file. So designers will be able to see exactly what is implemented, but from within their own set of tools, where they can then perform changes and add revisions.

Unfortunately, I found that the Figma API is a bit limited compared to what I had hoped for. Instead of being able to load a React component with form elements and various custom elements into Figma, it really only supports some basic shapes and constructs, such as `Rectangle`, `Ellipse`, etc.

But most web apps are not build from complex shapes. Most are just a lot of rectangles inside rectangles with a bit of images and svgs thrown in.

So why not try to build some sort of middleware between "production React components" and react-figma that could do stuff like this:

- Turn semantic HTML elements such as `div`, `section`, `article` into Figma `Rectangle` components.
- Parse as much of the CSS as possible and pass it as props on the Figma components.
- Build a rendering layer that calculates dimensions that are specified in relative terms (i.e. %) to absolute values (pixels). I imagine this could be done by running a headless browser and measuring element sizes. The user would have to specify certain preset widths for the components to be generated in (i.e. mobile 600px, tablet 960px, desktop 1280px for example)
- Potentially look into automatically outputting all components within a Storybook project to Figma in a structured way (perhaps a "Page" per component and a number of "Frames" representing the different preset component widths). This way the designer could have their own Storybook embedded into Figma. I imagine when a designer would want to do a new revision of a component they would copy the automatic generated component into a new design file, make changes and pass it on to the development team.

Is this something someone here is already working on? Any experience doing something like this? Is it a great idea or is it a terrible idea? Would you use a tool like this? Let me know what your thoughts are?

**Comment by ilyalesik** [comment]
I think it's interesting idea. Probably it would be better if it will be a separated package, such a `react-figma-web` or `react-figma-dom` that will contain a logic that will transform browser styles to 'primitive' styles (compared to usual react-figma API) by launching Puppeteer under the hood.
